feat: Bump Lucene@8.11.3 and lucene-gosen@8.11.0

miurahr opened the pull request:
Update Apache Lucene version from 5.5.5 to 8.11.3 that is latest version which support Java 8.
For dependency resolution, also update Lucene-Gosen 8.11.0

Unfortunately, Gosen project does not publish the version in Maven Central, I uses my build of Gosen.

Implement #3628 

## Changes

- regenerate ngram test data 
- Because `o.a.l.d.LongField` is removed in Lucene 8.x , switch to `LongPoint`, affected
        -  `languagetool-dev/src/main/java/org/languagetool/dev/bigdata/AggregatedNgramToLucene.java`
        -  `languagetool-dev/src/main/java/org/languagetool/dev/bigdata/CommonCrawlToNgram.java`
- `MultiFields.getFields(reader)` is deprecated, affected 
        -  `languagetool-dev/src/main/java/org/languagetool/dev/archive/StartTokenCounter.java` 
        - `languagetool-dev/src/main/java/org/languagetool/dev/bigdata/GermanUppercasePhraseFinder.java`
        - `languagetool-dev/src/main/java/org/languagetool/dev/bigdata/LargestNGramFinder.java`
        - `languagetool-dev/src/main/java/org/languagetool/dev/bigdata/NeededNGramCounter.java`
        - `languagetool-standalone/src/main/java/org/languagetool/dev/HomophoneOccurrenceDumper.java`
- `limitedTopDocs.topDocs.totalHits` to add `.value` that return long instead of int
-  Change `DirectoryReader.open(IndexWriter, boolean)` to `open(IndexWriter, boolean, boolean)` affected
       - `languagetool-dev/src/main/java/org/languagetool/dev/bigdata/CommonCrawlToNgram.java`
- `languagetool-wikipedia/pom.xml` change `lucene.version`
- `CharacterUtils` changes `fill` to be static function instead of a method

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of n-gram counts for improved indexing efficiency.
	- Improved term processing flexibility in the identification of uppercase phrases.

- **Bug Fixes**
	- Updated total hits access method to ensure consistency across the application.

- **Documentation**
	- Clarified import statements and dependencies in project configuration files.

- **Chores**
	- Updated multiple library dependencies to their latest versions for improved functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

miurahr committed:
feat: Bump Lucene@8.11.3 and lucene-gosen@8.11.0

miurahr committed:
fix: update for migrating lucene 8.11.3

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

miurahr committed:
chore: import lucene-core in -wikipedia

miurahr committed:
chore: bump lucene 8.11.3 for lt-wikiepdia

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

miurahr committed:
style: revert non-mandatory changes

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

miurahr commented:
You can check the change with upgraded English ngram data 
https://github.com/miurahr/lt-8-ngram-data

miurahr commented:
Here is an example of a usage of the patch.
This is a container configuration to launch `languagetool-server` that is built from source.
It uses a private forked version of LT with the modification proposed and Gradle build system.  
https://github.com/miurahr/languagetool-server-container

danielnaber commented:
Thanks!

> Unfortunately, Gosen project does not publish the version in Maven Central, I uses my build of Gosen.

Could you explain this in more details? Are all the dependencies still available at Maven Central?

miurahr commented:
> Thanks!
> 
> > Unfortunately, Gosen project does not publish the version in Maven Central, I uses my build of Gosen.
> 
> Could you explain this in more details? Are all the dependencies still available at Maven Central?

Sure,

Here is a project home of the forked lucene-gosen project.
https://github.com/omegat-org/lucene-gosen

The `rel-8.11` branch is to build lucene-gosen with Lucene 8.11.x.

You can find a `lucene-gosen` on Maven Central, please find it on https://mvnrepository.com/artifact/org.omegat.lucene/lucene-gosen

OmegaT project published three versions
- '5.0.0'   - we use it for OmegaT 6.0 with LanguageTool 3.5 and Lucene 5.2.1
- '5.5.1'  -  we use it for OmegaT 6.1 with LanguageTool 6.1 and Lucene 5.5.5
- '8.11.0' -  for the PR here, forked from the specific commit in the history, and built with Lucene 8.11.3.

miurahr commented:
> Are all the dependencies still available at Maven Central?

Yes, all the dependencies are available at Maven Central.

miurahr commented:
To upgrade other language ngram data to lucene 8 format,  https://github.com/hakanai/luceneupgrader will help.

danielnaber commented:
@miurahr Thanks again for your contribution. I've tested the PR on a live server, and everything looks good so far. The plan it to merge it shortly after the LT 6.5 release on Friday (we're in feature freeze, so we won't be able to merge it before).

miurahr committed:
Merge branch 'master' into topic/miurahr/migrate-lucene-8.11.3

miurahr commented on languagetool-wikipedia/src/main/java/org/languagetool/dev/index/Searcher.java:
```diff
@@ -200,7 +201,7 @@ public SearcherResult findRuleMatchesOnIndex(PatternRule rule, Language language
 
   private PossiblyLimitedTopDocs getTopDocs(Query query) throws IOException {
-    TopScoreDocCollector topCollector = TopScoreDocCollector.create(maxHits);
+    TopScoreDocCollector topCollector = TopScoreDocCollector.create(maxHits, Integer.MAX_VALUE);
```
@danielnaber This is the issue of design and policy of the library. Could you give me a number the limit?

danielnaber replied:
@miurahr As the ngram set is static, I assume we can ignore this advice here.

miurahr committed:
fix: adjust for review comment comment from CodeRabbbit

## language-dev sub-project

- Add StoredField addition to LongPoint for ngram
- Use DirectoryReader.open(IndexWriter) instead of open(IndexWriter, boolean, boolean)
- Null and empty check for FieldInfo and Terms

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

miurahr commented on languagetool-dev/src/main/java/org/languagetool/dev/bigdata/AggregatedNgramToLucene.java:
```diff
@@ -95,13 +95,8 @@ private Document getDoc(String ngram, long count) {
-    fieldType.setDocValuesType(DocValuesType.NUMERIC);
-    return new LongField("count", count, fieldType);
+  private LongPoint getCountField(long count) {
+    return new LongPoint("count", count);
```
I adjust it as suggested.

miurahr commented on languagetool-dev/src/main/java/org/languagetool/dev/bigdata/CommonCrawlToNgram.java:
```diff
@@ -177,16 +177,16 @@ private void writeToLucene(int ngramSize, Map<String, Long> ngramToCount) throws
       reader = newReader;
     }*/
-    index.reader = DirectoryReader.open(index.indexWriter, true);
+    index.reader = DirectoryReader.open(index.indexWriter, true, true);
```
Adjusted as suggested.

miurahr commented on languagetool-dev/src/main/java/org/languagetool/dev/bigdata/CommonCrawlToNgram.java:
```diff
@@ -269,7 +264,7 @@
       directory = FSDirectory.open(dir.toPath());
       indexWriter = new IndexWriter(directory, config);
-      reader = DirectoryReader.open(indexWriter, false);
+      reader = DirectoryReader.open(indexWriter, false, false);
```
Adjusted as suggested.

miurahr commented on languagetool-dev/src/main/java/org/languagetool/dev/bigdata/CommonCrawlToNgram.java:
```diff
@@ -221,13 +221,8 @@
-    fieldType.setDocValuesType(DocValuesType.NUMERIC);
-    return new LongField("count", count, fieldType);
+  private LongPoint getCountField(long count) {
+    return new LongPoint("count", count);
```
Adjusted as suggested.

miurahr commented on languagetool-dev/src/main/java/org/languagetool/dev/bigdata/GermanUppercasePhraseFinder.java:
```diff
@@ -57,47 +59,48 @@
-        if (part.length() < MIN_TERM_LEN) {
-          useful = false;
-          break;
+    for (String field: FieldInfos.getIndexedFields(reader)) {
```
A suggestion is not worked for me. I updated it differently.

miurahr commented on languagetool-dev/src/main/java/org/languagetool/dev/bigdata/GermanUppercasePhraseFinder.java:
```diff
@@ -57,47 +59,48 @@ public static void main(String[] args) throws IOException {
-          break;
+    for (String field: FieldInfos.getIndexedFields(reader)) {
+      Terms terms = MultiTerms.getTerms(reader, field);
+      TermsEnum termsEnum = terms.iterator();
```
Adjusted as suggested.

miurahr committed:
Merge branch 'master' into topic/miurahr/migrate-lucene-8.11.3

SteVio89 commented:
Sorry for the delay here. We're currently updating our backend to prepare everything for this PR.

miurahr commented:
ping

miurahr commented:
7 months anniversary from first proposal! @SteVio89 I hope you are well and have progress.

miurahr commented:
@danielnaber  LT 6.6 is out. Is there any plan for Lucene's upgrade?

SteVio89 commented:
@miurahr 

I'm deeply sorry this PR had to wait so long.
We'll revisit this topic in the upcoming weeks (definitely this quarter) and hopefully be able to complete it.

miurahr commented:
When the `languagetool` moves to Lucene 8.11.3 or later, there is a bonus.

Lucene 8.11 has a Hunspell implementation by pure java. The Hunspell native library is good for compatibility and functionality, but there is a difficulty to support new platform such as Windows 10/11 on Arm 64 and Linux on aarch64.
There are several merits to use Hunspell by Lucene; no need to sign for mac, and unlock platform dependency.

Here is a sample implementation of the idea.
https://github.com/miurahr/languagetool/pull/1

miurahr commented:
After several years, Lucene 8 has reached end-of-life. Additionally, the Apache Lucene Kuromoji analyzer has become the de facto standard tokenizer for Japanese, while lucene-gosen is no longer actively maintained.

I'd like to propose upgrading directly to Apache Lucene 10, which is actively maintained.

miurahr commented:
From Lucene 9, an index format has been changed. The maximum Lucene version is 9.12 to support index format 8, which is the version the PR here migrated.
